---
title: "Depression Microbiome Analysis"
author: "Cath Burke"
date: "123 August 2017"
output: html_document
---


```{r load packages}
library("phyloseq")
library("ggplot2")
library("vegan")
library("reshape2")
library("tidyverse")
```
#Load your data into Phyloseq
```{r Loading data}
#Open script in your working directory
#Set working directory, to source file location
#Create a folder called data in this directory - store all your raw data here
#first we assign names to filesbiom table must be
biom<-"data/otu_table_final1_json.biom"
#Biom table must be in json format
#Before reading the map file, remove the hash from the first line of the QIIME mapping file
map<-read.table("data/mapping_depression_f1.txt", header=TRUE, row.names=1)
#Normally would just use the read_tree command, but this particular tree is a bit different to ususal.
tree<-read_tree("data/rep_set.tre")
#import data to phyloseq
depression2017<-import_biom(biom, treefilename=tree)
depression_map<-sample_data(map)
depression2017<-merge_phyloseq(depression2017,depression_map)
depression2017
colnames(tax_table(depression2017)) = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
```
#Explore the data
```{r, explore data}
#look at elements of taxonomy table and otu table
#This shows the first 10 rows of the taxonomy table for your phyloseq object
tax_table(depression2017)[1:10]
#You can subset to a specific row by name - insert a row name in the ""
tax_table(depression2017)["",]
#this returns the first 10 rows and columns of the otu table
otu_table(depression2017)[1:10,1:10]
#plot data - generic plotting options in phyloseq.
#plot_bar(practice2017)
#plot_bar(practice2017,x="Disease_status")
#plot_bar(practice2017,x="Disease_status", fill="Phylum")
#You can also use ggplot for better plotting, but takes a bit of data wrangling, more on that below.

#Five number summary of sequence coverage of samples
fivenum(colSums(otu_table(depression2017)))
#View number of sequences per sample
seq_sums<-data.frame(sort(sample_sums(depression2017)))
#filter out rare OTUs - maybe only do this before statisitcal tests??
#depression2017_fil<- filter_taxa(depression2017, function(x) sum(x)>=10 & #sum(x>0)>0.25*length(x), TRUE)
```
#Check sequencing controls - mock communtiy
```{r,mock community}
#Look at blast results of mock seqs against reference seqs.  
mock_blast<-read.delim(file="data/mock_seqs_full_blast.txt", header=TRUE, sep="\t")

mock_blast %>% count(sseqid, pident) %>% 
  arrange(desc(pident))%>% arrange(sseqid) %>% group_by(sseqid) %>%
  dplyr::summarise(n_distinct(pident))
#Summarise the number of hits with a distinct percent identity for each taxa in the mock community
mock_summary<-mock_blast %>% group_by(sseqid, pident) %>%
  arrange(sseqid, desc(pident)) %>%
  summarise(n=n()) %>%
  arrange(sseqid, desc(pident)) %>% group_by(sseqid) %>%
  summarise(counts=sum(n))

mock_98<-mock_blast %>% 
  filter(pident>=98.000) %>%
  arrange(sseqid,desc(pident)) %>% group_by(sseqid, pident) %>%
  summarise(n=n()) %>%
  group_by(sseqid) %>%
  summarise(counts=sum(n))
  
mock_97<-mock_blast %>% 
  filter(pident>=97.000) %>%
  arrange(sseqid,desc(pident)) %>% group_by(sseqid, pident) %>%
  summarise(n=n()) %>%
  group_by(sseqid) %>%
  summarise(counts=sum(n))

#Calculate proportion of seqs with < 97% identity to reference
less_97<-mock_blast %>%
  filter(pident<97) %>%
  nrow()/nrow(mock_blast)
less_99<- mock_blast %>%
  filter(pident<99) %>%
  nrow()/nrow(mock_blast)  

#Calculate observed vs expected frequency of taxa for hits > 97% identity to a reference.
observed<- mock_97 %>% group_by(queryseqid,sseqid) %>%
  mutate(count=1)%>% group_by(sseqid) %>%
  summarise(counts=sum(count)) %>% mutate(observed=counts/sum(counts)*100)

expected<-c(15.7,10.4,10.0,18.8,15.9,4.6,11.3,13.3)

observed_expected<-cbind(observed,expected) %>% select(sseqid, observed, expected)
observed_expected$sseqid<-gsub("_16S[_2]*","", observed_expected$sseqid)
melt(observed_expected) %>% 
  ggplot(aes(x=sseqid,y=value, fill=variable))+geom_bar(stat="identity", position="dodge") + theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  labs(x="Mock community species", y="Relative abundance (%)",title="Observed vs Expected Frequencies of Species in the Mock Community")
```

#Rarefaction analysis
```{r, rarefaction}
#Create a rarefaction curve to find the appropraite place for rarefying your data. Uses vegan for calculating the curve, and base R for plotting. 
#Plot rarecurve
t_otu_depression2017<-as.matrix(t(otu_table(depression2017)))
#col <- c("forestgreen", "red", "hotpink", "black", "blue")
#cols<-col[sinus2017_data$Type]
curve<-rarecurve(t_otu_depression2017, step=5000, label=FALSE, xlim= c(0,50000), ylim=c(0,2000), ylab = "OTU Count", xaxt="n")+
axis(1, xaxp=c(0, 50000, 10))
#place a line over the curve where you proppose to rarefy, in this example I used 6000 sequences
abline(v=28170, col="red")
#Then rarefy at the appropriate depth - in this example I used 6000 sequences.
depression2017_rare<-rarefy_even_depth(depression2017, sample.size=28170, rngseed=711, replace=FALSE, trimOTUs=TRUE)
```
#subsetting phyloseq objects
```{r, subsetting phyloseq objects}
#You many want to remove certain samples or OTUs from the analysis.  There are several ways to do this in Phyloseq, all require creating a new Phyloseq object.
#Removed participant incorrectly enrolled
depression2017_rare_1<-subset_samples(depression2017_rare, !(SampleID %in% c("P9.pre","P7.pre","P7.post")))
```

#Alpha diversity analysis
```{r, alpha diversity}
#Once you have rarefied your data, explore the level of alpha diversity.  Uses phyloseq plot function which is built on ggplot2.  The x value will determine how samples are grouped along the x axis.
plot_richness(depression2017_rare_1, x="Disease_status", measures=c("Observed", "Chao1","Shannon"))+
  geom_boxplot()+xlab("Group")+ylab("Diversity")
#Convert alpha diversity estimates to a data form we can use for statistical testing
depression2017_rare_1_richness<-as.matrix(estimate_richness(depression2017_rare_1, measures=c("Observed","Chao1","Shannon")))
#row.names(practice2017_rare_richness)<-row.names(sample_data(practice2017_rare))
depression2017_rare_1_richness<-cbind(depression2017_rare_1_richness,sample_data(depression2017_rare_1))
depression2017_rare_1_richness
#For comaparisons of alpha diversity of more than two groups, use the Kruskal-Wallis test 
#Test if alpha diversity (observed) is significantly different between depressed_pre and controls
kruskal.test(Observed ~ Dep_Con, data=depression2017_rare_1_richness)

```
###UP TO HERE - DO EXAMINATION OF NEGATIVE AND POSITIVE CONTROLS
##Data cleaning and transformation
```{r data cleaning and transformation}
#Create an object with realtive abundances instead of counts
depression2017_rel_1<-transform_sample_counts(depression2017_rare_1, function(x) x/sum(x))
#Filter OTUs present at less that 0.001% mean abundance
depression2017_rel_2<-filter_taxa(depression2017_rel_1, function(x) mean(x) > 1e-5, TRUE)
```
##Plotting functions
```{r plotting}
#I use the psmelt function to get the data into a form that ggplot likes, dplyr to filter that data I want, and ggplot to do the plotting
#Melt data into a "long" form
depression_smelt<-psmelt(depression2017_rel_1)
head(depression_smelt)
#Look for Bacillus OTUs in data since it was lost from mock community.
bacillus_otus<-depression_smelt %>% filter(Genus == "g__Bacillus") %>%
  ggplot(aes(x=Sample,y=Abundance))+geom_bar(stat="identity")+facet_grid(.~Participant, scales="free")
  
fp_otus<-depression_smelt %>% filter(Species=="s__prausnitzii")

ggplot(fp_otus, aes(x=Sample,y=Abundance))+geom_bar(stat="identity")+facet_grid(.~Participant, scales="free")

ggplot(fp_otus, aes(x=Participant,y=Abundance))+geom_boxplot(stat = "boxplot")

fp_otus %>% group_by(SampleID) %>% mutate(sum_abund=sum(Abundance)) %>%
  distinct(SampleID, .keep_all=TRUE) %>%
  ggplot(aes(x=Participant,y=Abundance))+geom_boxplot(stat = "boxplot")

  


#Let's plot the overall taxa by group (average relative abundance per group) and colour by phyla
#plot every sample by phyla, faceting into disease groups, 
phyla_plot<-depression_smelt %>% group_by(SampleID,Phylum) %>%
  mutate(abund=sum(Abundance)) %>% distinct(Phylum, .keep_all=TRUE) %>% select(SampleID, Phylum, Abundance, abund, Participant) %>% arrange(SampleID,Phylum)

mycol<-c("#e6194b","#3cb44b","#ffe119","#0082c8","#f58231","#911eb4","#46f0f0","#f032e6","#d2f53c","#fabebe","#008080","#e6beff","#aa6e28","#fffac8","#800000","#aaffc3","#808000","#ffd8b1","#000080","#808080","#FFFFFF","#000000")

ggplot(phyla_plot, aes(x=SampleID, y=abund, fill=Phylum))+geom_bar(stat="identity")+facet_grid(.~Participant, scales="free")+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+ scale_color_brewer(palette="mycol")
#calculate the sum of abundance for each phyla per patient, then average out across groups and plot.
phyla_plot<-practice_smelt %>% group_by(Sample, Phylum, Disease_status) %>%
  summarise(phylum_sum=sum(Abundance))%>%
  group_by(Disease_status, Phylum) %>% summarise(phylum_av=mean(phylum_sum))
ggplot(phyla_plot, aes(x=Disease_status, y=phylum_av, fill=Phylum)) + geom_bar(stat="identity")+facet_grid(.~Disease_status, scales="free")+
  guides(fill=guide_legend(ncol=1))+theme(axis.text.x = element_text(angle = 90, hjust = 1),legend.key.size = unit(0.5,"line")) +ggtitle("Average Phyla per Group")

#Top ten OTUs per patient coloured by Genus
topten<-practice_smelt %>% group_by(Sample) %>% top_n(10,Abundance) %>%
  ggplot(aes(x=Sample, y=Abundance, fill=Genus))+geom_bar(stat="identity") +
  facet_grid(.~Disease_status, scales="free")+
  guides(fill=guide_legend(ncol=1))+ 
  theme(axis.text.x = element_text(angle = 90, hjust = 1),legend.key.size = unit(0.5,"line"))
  

```
##Beta-Diversity
```{r, beta-diversity}
#filter out rare OTUs
depression2017_1rare_fil<- filter_taxa(depression2017_1rare, function(x) sum(x)>=10 & sum(x>0)>0.1*length(x), TRUE)

depression2017_wuf<-distance(depression2017_1rare_fil, "wunifrac", type="samples")
#Warning message:
#In UniFrac(physeq, weighted = TRUE, ...) :
  #Randomly assigning root as -- 193336 -- in the phylogenetic tree in the data you provided.
depression2017_wuf_ord<-ordinate(depression2017_1rare_fil, method="PCoA", distance=depression2017_wuf)
plot_ordination(depression2017_1rare_fil, depression2017_wuf_ord, type="samples",color="Participant")

pre_con_rare_fil<-subset_samples(depression2017_1rare_fil, Participant %in% c("pre","Con"))
pre_con_ord<-ordinate(pre_con_rare_fil, method="PCoA", "wunifrac")
plot_ordination(pre_con_rare_fil,pre_con_ord, color="Participant")


```





## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
